app/uk/gov/hmrc/fileupload/virusscan/ScanningService.scala

```diff
@@ -28,23 +29,54 @@ object ScanningService {
   type ScanResult = Xor[ScanError, ScanResultFileClean.type]
 
   sealed trait ScanError
+
   case object ScanResultVirusDetected extends ScanError
-  case class  ScanResultFailureSendingChunks(ex: Throwable) extends ScanError
+
+  case object ScanReadCommandTimeOut extends ScanError
+
+  case class ScanResultFailureSendingChunks(ex: Throwable) extends ScanError
+
   case object ScanResultUnexpectedResult extends ScanError
-  case class  ScanResultError(ex: Throwable) extends ScanError
+
+  case class ScanResultError(ex: Throwable) extends ScanError
+
   case object ScanResultFileClean
 
   type AvScanIteratee = Iteratee[Array[Byte], Future[ScanResult]]
 
   def scanBinaryData(scanner: () => Iteratee[Array[Byte], Future[ScanResult]],
+                     scanTimeoutAttempts: Int,
                      getFile: (String, String) => Future[QuarantineDownloadResult])
                     (s3KeyAppender: (EnvelopeId, FileId) => String)
-                    (envelopeId: EnvelopeId, fileId : FileId, fileRefId: FileRefId)
+                    (envelopeId: EnvelopeId, fileId: FileId, fileRefId: FileRefId)
                     (implicit ec: ExecutionContext): Future[ScanResult] = {
     val appendedKey = s3KeyAppender(envelopeId, fileId)
-    getFile(appendedKey, fileRefId.value).flatMap {
+    retries(scan(scanner, getFile(appendedKey, fileRefId.value)), scanTimeoutAttempts, fileId)
+  }
+
+  private def retries(scanResult: => Future[ScanResult],
+                      timeoutAttempts: Int,
+                      fileId: FileId,
+                      acc: Int = 1)
+                     (implicit ec: ExecutionContext): Future[ScanResult] = {
+    scanResult.flatMap {
+      case result if acc == timeoutAttempts =>
```

Comment:
As a side-note, I was having slight troubles understanding what happens if I put `numberOfTimeoutAttempts = 0` in the config, just in case we wanted to disable retries.

I *think* it might be fine, as this line in the ScanningService would be 0 == 1 ... and then it would just continue to the default case in the match. I couldn't spot a test for that though - did I miss it, and/or was it something you considered?

Reply:
You're correct the last pattern match would kick in and the retry logic would not take effect. It would behave exactly the same way for negative numbers too. 

numberOfTimeoutAttempts in config is optional. When it isn't there, we default to 1 so these really are edge cases, but quick enough to write tests for. 👍

Reply:
Also amended so that's >= rather than ==, in case config is ever set to zero.